src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -0,0 +1,212 @@
+import { html, LitElement } from "@polymer/lit-element";
+import "@polymer/paper-checkbox/paper-checkbox.js";
+import "@polymer/paper-icon-button/paper-icon-button.js";
+import "@polymer/paper-input/paper-input.js";
+import "@polymer/paper-item/paper-icon-item.js";
+import "@polymer/paper-item/paper-item-body.js";
+
+import "../../../components/ha-card.js";
+
+import { HassLocalizeLitMixin } from "../../../mixins/lit-localize-mixin";
+import { HomeAssistant } from "../../../types.js";
+import { LovelaceCard, LovelaceConfig } from "../types.js";
+import { clearItems, fetchItems, addItem, completeItem, saveEdit } from "../../../data/shopping_list";
+
+interface Config extends LovelaceConfig {
+  title?: string;
+}
+
+interface ListItem {
+  id: number;
+  name: string;
+  complete: boolean;
+}
+
+class HuiShoppingListCard extends HassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  protected hass?: HomeAssistant;
+  protected config?: Config;
+  private items?: ListItem[];
+
+  static get properties() {
+    return {
+      hass: {},
+      config: {},
+      items: [],
+    };
+  }
+
+  public getCardSize() {
+    return (this.config ? (this.config.title ? 1 : 0) : 0) + (this.items ? this.items.length : 0);
+  }
+
+  public setConfig(config: Config) {
+    this.config = config;
+    this.items = [];
+    this._fetchData();
+  }
+
+  public connectedCallback() {
+    if (!this.hass) {
+      return;
+    }
+
+    super.connectedCallback();
+    this.hass.connection
+      .subscribeEvents(this._fetchData, "shopping_list_updated")
```

Comment:
subscribeEvents resolves to an unsub function. The card should unsub when it is disconnected

Reply:
I'm still lost as how this should work

Reply:
Here's the code from `ha-panel-shopping-list.js`
```
  connectedCallback() {
    super.connectedCallback();
    this._fetchData = this._fetchData.bind(this);

    this.hass.connection
      .subscribeEvents(this._fetchData, "shopping_list_updated")
      .then(
        function(unsub) {
          this._unsubEvents = unsub;
        }.bind(this)
      );
    this._fetchData();
  }

  disconnectedCallback() {
    super.disconnectedCallback();
    if (this._unsubEvents) this._unsubEvents();
  }
```

I don't understand the `this._unsubEvents` portion

Reply:
So when we subscribe to events, we create a subscription for the current websocket connection. The backend will keep sending the events over the websocket connection as long as the subscription is around. So when we no longer need this component and remove it from the DOM (and we're notified by `disconnectedCallback` being called), we should remove that subscription, or else the backend will keep sending updates. Now imagine yoy refresh your lovelace config 10 times, we now have 10 subscriptions…